### DISCUSSION TITLE
Preserve values across test re-runs in Watch mode

### DISCUSSION THREAD

**Discussion by anthumchris** [avajs/ava#2803] [Q&A] [answered]
How can we persist values across test runs that occur in Watch mode? For example, I'd like to
1. run Ava in watch mode
2. instantiate and store a value when Ava runs test for the first time
3. modify a test file
    1. (Ava re-runs the test file)
5. The test will fetch the pre-existing value.

My use case is creating a pool of [Puppeteer](https://github.com/puppeteer/puppeteer) pages that can quickly be obtained and worked with.  These pages take > 5s to authenticate and load, and I'd like the Puppeteer browser and created pages to be reused to avoid the instantiation times.

**Comment by novemberborn** [comment, marked as answer]
This would be possible using the experimental [shared workers](https://github.com/avajs/ava/blob/main/docs/recipes/shared-workers.md) feature.

Keep in mind though that test files run in fresh contexts, so you won't be able to directly share the Puppeteer API.

**Comment by anthumchris** [comment]
I will look into the shared workers. My current workaround is to launch Chrome with a debugging port and use `puppeteer.connect()` to connect to that.  This allows Chrome to stay open and maintain existing page authentications.

```sh
$ "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome" --remote-debugging-port=9222
```

```js
import puppeteer from 'puppeteer-core'

puppeteer.connect({
  browserURL: 'http://127.0.0.1:9222/json/version',
  ...
})
```

**Reply by novemberborn** [reply]
That looks an approach that would work with shared workers as well.
